### DISCUSSION TITLE
Persist middleware causing typescript error with boolean values

### DISCUSSION THREAD

**Discussion by johnslemmer** [pmndrs/zustand#2231] [Bug report]
## Summary

This is certainly related to https://github.com/pmndrs/zustand/discussions/2068 and https://github.com/pmndrs/zustand/discussions/2163 which point to the merged PR https://github.com/pmndrs/zustand/pull/2170 suggesting that this should be fixed.  But it isn't.  There was also some mention of `tsconfig.moduleResolution` being set to `bundler` causing issues.  But both on my linked codesandbox and my local project that is not the case. So I'm starting a new discussion.

Here is an example of the code that errors.  This seems unexpected.

```js
import { create } from "zustand";
import { persist } from "zustand/middleware";

type FooStore = {
  foo: string;
};

type BarStore = {
  bar: boolean;
};

// NOTE: typescript does not error on these
const useFooStore = create<FooStore>()(() => ({
  foo: "asdf"
}));
const useFooStorePersist = create<FooStore>()(
  persist(
    () => ({
      foo: "asdf"
    }),
    { name: "foo" }
  )
);
const useBarStore = create<BarStore>()(() => ({
  bar: false
}));

// NOTE: typescript does error here
const useBarStorePersist = create<BarStore>()(
  persist(
    () => ({
      bar: false
    }),
    {
      name: "bar"
    }
  )
);
```

Error being generated: `Type 'boolean' is not assignable to type 'false'.ts(2345)`

Certainly there is a workaround: `loading: false as boolean` but that seems like that shouldn't be needed. And per the [documentation](https://docs.pmnd.rs/zustand/guides/typescript#using-middlewares) "You do not have to do anything special to use middlewares in TypeScript."

Tried this on versions 4.3.9, 4.4.6, 4.4.7. Typescript 5.3.2 and 5.3.3.  Node 20.10.0.

Any ideas?

## Link to reproduction

https://codesandbox.io/s/romantic-williamson-3hz8vl

(try the old editor on codesandbox, the new editor is being real buggy today)

**Comment by dai-shi** [comment]
https://github.com/pmndrs/zustand/discussions/2068#discussioncomment-7093074

TS Playground: https://tsplay.dev/Wo1O8N

I don't think there's anything we can do. Would anyone like to create a reproduction without Zustand and report it to https://github.com/microsoft/TypeScript/issues?

**Reply by johnslemmer** [reply]
Ok, thanks for changing the answer on #2068 that makes things more clear.

I'll let my brain stew on what is happening underneath the hood of zustand type-wise to see if a simpler typescript only reproduction comes to mind.
